### PR TITLE
Fix github actions for large file storage

### PR DESCRIPTION
**Pull request by CunliangGeng**
We used large file storage (lfs) for handling large test dataset (zip files). 

The github action " [actions/checkout]( https://github.com/actions/checkout)" does not use lfs automatically, leading to failed tests, and so we have to set it explicitly (see the [usage](https://github.com/actions/checkout)).

However, it's not free to use lfs and there is storage and bandwidth limit (see [the comment](https://github.com/NPLinker/nplinker/pull/224#issuecomment-1996709712)), which we do not want.

This PR stopped the use of lfs and implemented an alternative way to handle large files. The approach is to put the large files in Zenodo, and download it before running tests if there is no local cache.

Major changes:
- Stopped the use of lfs ([instructions](https://gist.github.com/everttrollip/198ed9a09bba45d2663ccac99e662201))
- Separated tests to unit tests and integration tests
- Uploaded large test dataset to Zenodo
- Implemented code to prepare test datasets and folders before running tests
- Excluded whole tests folder for distribution (i.e. pypi package)

### CODE REVIEW

**Comment by CunliangGeng** [comment]
Got an email from github on warning of the bandwidth usage of lfs (see below), it looks lfs is not a great solution for us. So I converted this PR to draft and will explore other solutions.

```
You’ve used 100% of your data plan for Git LFS on the organization NPLinker. Please purchase additional data packs to cover your bandwidth and storage usage:

  https://github.com/organizations/NPLinker/billing/data/upgrade

Current usage as of 13 Mar 2024 06:50PM UTC:

  Bandwidth: 1.04 GB / 1 GB (104%)
  Storage: 0.21 GB / 1 GB (21%)
```

File: pyproject.toml [hunk 1]
```diff
@@ -48,8 +48,8 @@ namespaces = true    # enable data directory to be identified
 
 [tool.pytest.ini_options]
```
Comment (gcroci2) [inline pyproject.toml:49]: You mentioned that with this PR you excluded the whole tests folder for distribution (i.e. pypi package), but I don't see any change in the toml in this regard. I was expecting something like:
```
[tool.setuptools.packages.find]
include = ["nplinker*"]
exclude = ["tests", "tests*", "*tests.*", "*tests"]
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -34,7 +34,7 @@ jobs:
     runs-on: ubuntu-latest
     if: github.event.pull_request.draft == false
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
```
Comment (gcroci2) [inline .github/workflows/build.yml:37]: Why is this upgrade to v4 needed?

File: MANIFEST.in [hunk 3]
```diff
@@ -3,3 +3,4 @@ include LICENSE
 include NOTICE
 include requirements.txt
 include requirements.dev.txt
+prune tests
```
Comment (gcroci2) [inline MANIFEST.in:6]: What is this for?

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+# The temporary directory for the test session
+temp_dir = tempfile.gettempdir()
+nplinker_root_dir = os.path.join(temp_dir, "nplinker_local_mode_example")
+
+
+def pytest_sessionstart(session):
+    """Pytest hook to run before the entire test session starts.
+
+    This hook makes sure the temporary directory `nplinker_root_dir` is created before any test
+    starts. When running tests in parallel, the creation operation is done by the master processs,
```
Comment (gcroci2) [inline tests/integration/conftest.py]:
```suggestion
    starts. When running tests in parallel, the creation operation is done by the master process,
```

File: tests/integration/conftest.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+dataset_url = (
+    f"https://zenodo.org/records/{dataset_doi.split('.')[-1]}/files/nplinker_local_mode_example.zip"
+)
+
+# The temporary directory for the test session
+temp_dir = tempfile.gettempdir()
+nplinker_root_dir = os.path.join(temp_dir, "nplinker_local_mode_example")
+
+
+def pytest_sessionstart(session):
```
Comment (gcroci2) [inline tests/integration/conftest.py:25]: Where is the `Session` object defined? [Here](https://docs.pytest.org/en/stable/reference.html#_pytest.hookspec.pytest_sessionstart) I didn't find examples of how to use initialization hooks.

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    # # Specify the config file via environment variable before importing nplinker in any test.
+    os.environ["NPLINKER_CONFIG_FILE"] = str(DATA_DIR / "nplinker_local_mode.toml")
+
+
+def pytest_sessionfinish(session):
+    """Pytest hook to run after the entire test session finishes.
+
+    This hook makes sure that temporary directory `nplinker_root_dir` is only removed after all
+    tests finish. When running tests in parallel, the deletion operation is done by the master
+    processs, and worker processes are not allowed to do it.
```
Comment (gcroci2) [inline tests/integration/conftest.py]:
```suggestion
    process, and worker processes are not allowed to do it.
```

File: tests/unit/conftest.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+
+nplinker_root_dir = os.path.join(tempfile.gettempdir(), "nplinker_unit_test")
+
+
+def pytest_sessionstart(session):
+    """Pytest hook to run before the entire test session starts.
+
+    This hook makes sure the temporary directory `nplinker_root_dir` is created before any test
+    starts. When running tests in parallel, the creation operation is done by the master processs,
```
Comment (gcroci2) [inline tests/unit/conftest.py]:
```suggestion
    starts. When running tests in parallel, the creation operation is done by the master process,
```

**Review by gcroci2 — APPROVED** [review]
Great PR! Only minor comments and suggestions.

File: pyproject.toml (hunk 1 above)
Comment (CunliangGeng) [reply pyproject.toml:49]: It's managed by the file `MANIFEST.in`.

File: MANIFEST.in (hunk 3 above)
Comment (CunliangGeng) [reply MANIFEST.in:6]: This `MANIFEST.in` file manages what to include and exclude for pypi packaging. So `prune tests` means exclude the whole folder of `tests`.

File: .github/workflows/build.yml (hunk 2 above)
Comment (CunliangGeng) [reply .github/workflows/build.yml:37]: v4 is the latest version, so I just upgraded it. v3 works also well.

File: tests/integration/conftest.py (hunk 5 above)
Comment (CunliangGeng) [reply tests/integration/conftest.py:25]: I did not look into how the hook `pytest_sessionstart(session)` works in pytest, but it's the way how to use it. If you want to take a look, check here https://docs.pytest.org/en/stable/reference/reference.html#pytest.hookspec.pytest_sessionstart.

**Comment by CunliangGeng** [comment]
* **#229** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#228** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#227** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#226** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#224** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/224?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `dev`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @CunliangGeng and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by CunliangGeng** [comment]
### Merge activity

* **Mar 21, 7:58 AM EDT**: @CunliangGeng started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/224).
* **Mar 21, 7:59 AM EDT**: @CunliangGeng merged this pull request with [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/224).

**Pull request merged**